### ISSUE TITLE
Some inappropriate characters are included in auto-generated permalinks

### ISSUE THREAD

**Issue by greatislander** [WordPress/gutenberg#12907]
**Describe the bug**

In WordPress 5.0.1, creating a post or page with colons or exclamation marks in the title produces a permalink that takes the expected post/page slug form (lowercase alphanumeric with hyphens instead of spaces) but includes these special characters. Once the post/page is published the characters are stripped from the permalink.

**To Reproduce**

1. Create a new post or page.
2. Enter a title with colons or exclamation marks.
3. Save draft.
4. Inspect the permalink.

**Expected behavior**

The permalink should contain only alphanumeric characters and hyphens, even before publishing the post/page.

**Screenshots**

The bug visible in the page title:

<img width="614" alt="Screen shot of the page title and permalink." src="https://user-images.githubusercontent.com/605361/50044675-9697cc80-005d-11e9-9606-38a622c6ff6e.png">

The bug visible in the document inspector:

<img width="278" alt="Screen shot of the permalink the document inspector." src="https://user-images.githubusercontent.com/605361/50044682-bcbd6c80-005d-11e9-8e38-b0d876c7e287.png">

**Desktop:**

 - OS: macOS Mojave 10.14.2
 - Safari 12.0.2

**Additional context**

- I'm using the WordPress 5.0.1 bundled version of Gutenberg.
- Other characters may be included as well, but I ran into the issue with these two (colon and exclamation mark) and haven't explored further.

**Comment by swissspidy** [comment]
Core ticket: https://core.trac.wordpress.org/ticket/45588

**Comment by Wolfneo** [comment]
Hi,
Same problem, thank you for finding the solution, thank you, in advance of your work.
Thank you

**Comment by apermo** [comment]
Same problem here, including that umlaute are displayed wrong as long as the post is a draft.

So ä becomes a, instead of ae, as it should. Once its saved, everything is ok.

**Comment by DominicDesbiens** [comment]
I searched for this issue in WordPress with Gutenberg. Found some another tickets about this problem. Is there any news about this issue ? Still there with WordPress 5.1.1.

Thanks

**Comment by khoipro** [comment]
Could we reopen this ticket? Our Vietnamese language still has an issue with that slug. @swissspidy @earnjam @ocean90 

Track my ticket https://core.trac.wordpress.org/ticket/54492

Thanks, @SergeyBiryukov for referring here.

**Comment by IIM-Arvid** [comment]
Hi,

thanks for reopening the ticket. There are still problems mit other languages when saving the post as a draft. As stated in #16695 by @uok german "Umlauts" are not treaded properly (tested in 5.9.1).
When creating a new post and entering a title like "Am nächsten Tag" and saving the post as a draft, the JS function mentioned above ( cleanForSlug() ) creates the slug "am-nachsten-tag" which is wrong. It has to be "am-naechsten-tag".

If you save the post with other post-stati like "publish" or "future" the right slug is being created. 
The problem is that our editorial team is working with drafts to share the posts with other team members etc. They also use the generated slug / permalink for other purposes. For example: They copy the permalink and enter it in a browser push service. At the point the browser push is send out to thousands of users, the permalink has changed already because the post-status was now changed to "publish" and therefore the right slug was generated. Now all the users that click on the push message will see a 404 error page because the url was not found. And there are other cases in which this behavior causes problems.

Since sanitize_title() is working like a charm on the PHP side, i have created this "light" workaround in our own plugin until the core JS function does handle this as well:

add_filter( 'wp_insert_post_data', array( $this, 'sanitizeDraftPermalink' ) );

public function sanitizeDraftPermalink( array $postData ) : array
{
    if ( $postData['post_status'] === 'draft' && empty( $postData['post_name'] ) ) {
        $postData['post_name'] = sanitize_title( $postData['post_title'] );
    }
    return $postData;
}

Of course the JS side is not affected by this. So if you save the post as a draft, the wrong slug is still being created and shown. But if you reload the post now, the right slug is shown. At least that helps a little bit.

Maybe there is a way to adopt the behavior from sanitize_title() on the JS side.

Thank you very much

**Comment by Steve-Fenton** [comment]
Hello, I notice this item was closed in favour of this issue: https://core.trac.wordpress.org/ticket/47532

I just wanted to see if the "Save Draft" issue generating dash-runs was in the general awareness?

"example - example" becomes "example---example" when you save a draft, but this isn't what happens if you publish before saving a draft (it should be "example-example").

**Comment by Mamaduka** [comment]
Hi, @Steve-Fenton

I created PR to fix the issue with multiple hyphens - https://github.com/WordPress/gutenberg/pull/44873.

I can also reproduce the issue reported by @khoipro. I will try to follow up on that bug shortly.

**Comment by anubisthejackle** [comment]
> Maybe there is a way to adopt the behavior from sanitize_title() on the JS side.

What I've done to work around this on projects that I'm involved in is to create an endpoint that passes the draft title to `sanitize_title` and returns the slug. This is a relatively lightweight endpoint, and is only used during the editing process, so--while not as efficient as a fully frontend solution--it does the job well enough.

**Comment by Mamaduka** [comment]
Pinging @tyxla since he maintains the new library we're using for removing accents. I think the remaining issue is to have the same diacritics coverage on the JS side as we have in PHP

**Comment by tyxla** [comment]
Thanks for the ping @Mamaduka!

I took a look at the conversation above, and did some testing, and I was unable to reproduce any of the following bugs reported above:

* Multiple hyphens in a slug (fixed in https://github.com/WordPress/gutenberg/pull/44873)
* Colons / exclamation marks being included in slugs.

In general `remove-accents` is meant to be a clone of the WP `remove_accents()` function, which it is, but we did a few additional improvements over the years. I'm happy to iterate, but I really need to understand what the problem is first. Anyone got a clue or the right steps to reproduce the issue?

**Comment by Mamaduka** [comment]
@tyxla, I believe the remaining issue is with Vietnamese diacritics, initially reported here - https://core.trac.wordpress.org/ticket/54492.

## Steps to reproduce

1. Open post editor.
2. Run the following snippet in the console - `wp.url.cleanForSlug('Thiết kế nội thất');`
3. Notice that not all accents are removed.

**Comment by tyxla** [comment]
Cool, thank you @Mamaduka! I'll work on addressing those.

**Comment by swissspidy** [comment]
@tyxla Just curious, I assume you're able to handle those in your library? Looks like the version on GitHub is not the same as the one on npm

**Comment by tyxla** [comment]
> Just curious, I assume you're able to handle those in your library? 

Yes. 

> Looks like the version on GitHub is not the same as the one on npm

Oops, seems like we forgot to tag the release here and only did it in npm. Just published it on GitHub as well: https://github.com/tyxla/remove-accents/releases/tag/v0.4.3. Thanks, @swissspidy!

**Comment by tyxla** [comment]
@khoipro could I ask you to send me a thorough list of all the accented characters? 

My initial research shows that there are a bunch of different combinations of diacritics in every single character and I want to make sure I'm not missing any. Thanks in advance!

**Comment by khoipro** [comment]
@tyxla @SergeyBiryukov I bump this ticket again.

Here is the example of how to sanitize Vietnamese characters slug by another guy, but it works: https://gist.github.com/khoatran/8afeccd7a44c52ad39cfee00a093ec31

Another case as old plugin:

https://github.com/wp-plugins/vietnamese-slug/blob/master/vietnamese_slug.php

**Comment by tyxla** [comment]
@khoipro - thanks, I'm happy to work on a new version of `remove-accents` that includes all those strings. Stay tuned.

**Comment by tyxla** [comment]
Vietnamese support was added in the latest version of `remove-accents` and I'm bumping its version and adding a test in #53420. From what I understand once we land it, we can close this issue.

@khoipro let me know if that works well for you.

**Comment by tyxla** [comment]
After shipping #53420 I've closed this one, but if there's any additional character needing support, feel free to reopen or open a new issue and I'd be happy to take care of it!

**Comment by apermo** [comment]
Reopening the German umlaut issue, as this is still present and has a real data-integrity impact that wasn't fully described before.
 
 ### The core problem 
 
 German umlauts require locale-specific digraph replacements, not simple diacritic removal: 

 - `ä` → `ae,` `ö` → `oe`, `ü` → `ue`, `ß` → `ss` (correct, German/de_DE)
 - `ä` → `a`, `ö` → `o`, `ü` → `u`, `ß` → `s` (what `cleanForSlug()` currently produces)

 This is not a missing character mapping — it's a locale rule that only applies in German context. PHP's `remove_accents()` handles this via a de_DE locale check. The remove-accents npm package has no locale concept, so it always produces the non-German result.

 Steps to reproduce — cosmetic only (wrong preview)

 1. Set site language to de_DE (Settings → General → Site Language)
 2. Open the block editor, create a new post
 3. Enter title: `Künstler überraschen Hörer`
 4. Observe slug preview in the editor: `kunstler-uberraschen-horer`
 5. Publish without touching the slug
 6. Observe actual saved permalink: `kuenstler-ueberraschen-hoerer` ✓

 Server-side `sanitize_title()` generates the correct slug on publish because `post_name` is empty and PHP uses the site locale. So far, cosmetic only.

 Steps to reproduce — real data integrity bug

 1. Same setup, steps 1–4 above
 2. Observe that the editor shows `kunstler-uberraschen-horer`
 3. Click the slug to edit it — normal SEO workflow, user wants to shorten it
 4. The slug input is pre-filled with the wrong JS-generated value
 5. User types a shortened version based on what they see: `kunstler-horer` (missing the `ue` replacements)
 6. Publish
 7. Saved permalink: `kunstler-horer` — permanently wrong

 Once a user manually edits the slug, WordPress locks it — subsequent title changes no longer update it. The wrong transliteration is now permanent and the post has the wrong URL for its entire lifetime.

### Why this affects editorial teams specifically

 Our editors routinely customize slugs for SEO. The workflow is: write title → check auto-generated slug → shorten/optimize → publish. Step 2 of that workflow now shows them a wrong value, and any edit they make at step 3 bakes that wrong value in.

 ###Possible fix

 The cleanest solution would be for the editor to use the `generated_slug` field already returned by the REST API (computed server-side via `sanitize_title()`) to populate/correct the slug preview, rather than relying solely on client-side `cleanForSlug()`. The server already does the right thing — the editor just isn't using its output.

 Alternatively, locale-specific rules could be added to `cleanForSlug()` / remove-accents, using the site locale available from the REST API settings endpoint.

Disclaimer: Research and documentation supported by Claude Code and results verified by myself.

Edit: @tyxla I can't reopen it, missing the rights to it, would you be so kind to reopen it?

**Comment by tyxla** [comment]
This is an interesting problem, thanks @apermo, reopening!
